prompt2model/model_trainer/peft_trainer.py

```diff
@@ -0,0 +1,243 @@
+from accelerate import Accelerator, FullyShardedDataParallelPlugin
+from peft import LoraConfig, PeftModel, get_peft_model, prepare_model_for_kbit_training
+from torch.distributed.fsdp.fully_sharded_data_parallel import (
+    FullOptimStateDictConfig,
+    FullStateDictConfig,
+)
+from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig
+
+fsdp_plugin = FullyShardedDataParallelPlugin(
+    state_dict_config=FullStateDictConfig(offload_to_cpu=True, rank0_only=False),
+    optim_state_dict_config=FullOptimStateDictConfig(
+        offload_to_cpu=True, rank0_only=False
+    ),
+)
+
+accelerator = Accelerator(fsdp_plugin=fsdp_plugin)
+
+
+class EvalAccuracyCallback(transformers.TrainerCallback):
+    """A callback to evaluate the model on the validation set and log the accuracy."""
+
+    def __init__(self, val_data: datasets.Dataset, eval_tokenizer) -> None:
+        """Initialize the callback with the validation data and tokenizer."""
+        range_len = min(100, len(val_data))
+        self.val_data = val_data.shuffle(seed=42).select(range(range_len))
+        self.eval_tokenizer = eval_tokenizer
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    def on_evaluate(self, args, state, control, **kwargs):
+        """Evaluate the model on the validation set and log the accuracy."""
+        # Evaluate the model on the validation set
+        # 1. use the val data
+        # 2. use model and tokenizer
+        # 3. get accuracy
+        model = kwargs.get("model")
+        metrics = kwargs.get("metrics")
+        val_acc = 0
+        with torch.no_grad():
+            for row in self.val_data:
+                input_text = row["input_col"]
+                target = row["output_col"]
+
+                max_gen_len = (
+                    len(self.eval_tokenizer(target, return_tensors="pt").input_ids[0])
+                    + 5
+                )
+                inpt_ids = self.eval_tokenizer(
+                    input_text, return_tensors="pt"
+                ).input_ids.to(self.device)
+                generated_ids = model.generate(
+                    input_ids=inpt_ids,
+                    max_new_tokens=max_gen_len,
+                    do_sample=False,
+                    temperature=0,
+                )
+                text = self.eval_tokenizer.decode(
+                    generated_ids[0], skip_special_tokens=True
+                )
+                text = text.replace(input_text, "").lower().strip()
+                target = target.lower().strip()
+                if target in text:
+                    val_acc += 1
+        val_acc = val_acc / len(self.val_data)
+        print(f"Validation accuracy: {val_acc}")
+        if metrics is not None:
+            metrics["eval_accuracy"] = val_acc
+
+
+class QLoraTrainer:
+    """A class for fine-tuning a model using QLora."""
+
+    def __init__(self, model_name="mistralai/Mistral-7B-v0.1", eval_size=50) -> None:
```

Comment:
```suggestion
    def __init__(self, model_name: str, eval_size: int=50) -> None:
```

It'd probably be better to not have a default model name and just force people to specify it.